### PR TITLE
Fix feature indices

### PR DESCRIPTION
**Pull request by WilliamHeuser**
This PR removes the feature_indices from the .fit() function of a DecisionTree, and instead implements max_features where a subset, of size max_features, from the features are chosen at random for each split.

I profiled this and saw that the __get_feature_indices() method I implemented to help implement max_features ran for 2 milliseconds in a run of about 7500 milliseconds. Thus the time spent doing this is miniscule. 
I don't think that it is a problem for us with this added running time.

### CODE REVIEW

File: src/adaXT/decision_tree/DecisionTree.pyx [hunk 1]
```diff
@@ -44,6 +46,41 @@ class DecisionTree:
...
+    def __error_check_max_features(self, max_features):
+        if max_features is None:
+            return max_features
+        elif isinstance(max_features, int):
+            if max_features < 1:
+                raise ValueError("max_features can not be less than 1")
+            else:
+                return max_features
+        elif isinstance(max_features, float):
+            return max_features
```
Comment (svbrodersen) [inline src/adaXT/decision_tree/DecisionTree.pyx:58]: Should we still not make sure it is non negative even when it is a float? Perhaps even check
that the value of the float is between 0 and 1.

File: src/adaXT/decision_tree/DepthTreeBuilder.py [hunk 2]
```diff
@@ -88,6 +88,10 @@ def __init__(
         else:
             self.splitter = Splitter(self.features, self.response, criteria)
 
+        _, col = X.shape
+        self.feature_indices = np.arange(col, dtype=np.int32)
+        self.num_features = col
```
Comment (svbrodersen) [inline src/adaXT/decision_tree/DepthTreeBuilder.py:93]: Should probably be changed to num_tot_features, as the max_features could be less than this value.

**Review by svbrodersen — CHANGES_REQUESTED** [review]

File: src/adaXT/decision_tree/DecisionTree.pyx [hunk 3]
```diff
@@ -44,6 +46,41 @@ class DecisionTree:
         self.classes = None
         self.splitter = splitter
 
+    def __error_check_max_features(self, max_features):
```
[the remark this answers is not in the document]
Comment (WilliamHeuser) [reply src/adaXT/decision_tree/DecisionTree.pyx:49]: Done.

File: src/adaXT/decision_tree/DecisionTree.pyx (hunk 1 above)
Comment (WilliamHeuser) [reply src/adaXT/decision_tree/DecisionTree.pyx:58]: Done.

File: src/adaXT/decision_tree/DepthTreeBuilder.py (hunk 2 above)
Comment (WilliamHeuser) [reply src/adaXT/decision_tree/DepthTreeBuilder.py:93]: Done.

**Comment by NiklasPfister** [comment]
@svbrodersen: Once/if you agree with the changes (in particular the comment you had) you can merge this.

File: tests/test_tree_features.py [hunk 4]
```diff
@@ -200,7 +200,61 @@ def test_min_samples_split_setting():
...
+            is_correct = True
+        except BaseException:
+            is_correct = False
+            break
+
+    assert is_correct, "a non ValueError occured; when passing illegal arguments to max_features only a ValueError should be thrown"
+
+
+def test_max_features_setting():
+    test_max_features_errors()
```
Comment (svbrodersen) [inline tests/test_tree_features.py:225]: There is no reason to call the features errors here. Pytest will run it itself.

**Review by svbrodersen — CHANGES_REQUESTED** [review]

**Review by svbrodersen — APPROVED** [review]

**Pull request merged**